bmap-rs/src/main.rs

```diff
@@ -74,53 +83,150 @@ impl SeekForward for Decoder {
+        Input::Local(_) => find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?,
+        Input::Remote(_) => {
+            let mut url_bmap = c.image.clone();
+            url_bmap.set_extension("bmap");
+            url_bmap
+        }
+    };
 
-    let bmap = find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?;
     println!("Found bmap file: {}", bmap.display());
 
-    let mut b = File::open(&bmap).context("Failed to open bmap file")?;
-    let mut xml = String::new();
-    b.read_to_string(&mut xml)?;
-
-    let bmap = Bmap::from_xml(&xml)?;
-    let mut output = std::fs::OpenOptions::new()
-        .write(true)
-        .create(true)
-        .open(c.dest)?;
+    let bmap = match input_type {
+        Input::Local(_) => {
+            let mut b = File::open(&bmap).context("Failed to open bmap file")?;
+            let mut xml = String::new();
+            b.read_to_string(&mut xml)?;
+            Bmap::from_xml(&xml)?
+        }
 
-    if output.metadata()?.is_file() {
-        ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
-            .context("Failed to truncate file")?;
+        Input::Remote(_) => {
+            let url_bmap = bmap
+                .to_str()
+                .expect("Fail to convert remote path to str")
+                .parse::<Uri>()
+                .expect("Fail to convert remote path to url");
+            let mut bmap_res = match url_bmap.scheme_str() {
+                Some("https") => fetch_url_https(url_bmap).await?,
+                Some("http") => fetch_url_http(url_bmap).await?,
+                _ => {
+                    bail!("url is not http or https")
+                }
+            };
+            let mut c = Cursor::new(Vec::new());
+            while let Some(next) = bmap_res.data().await {
+                let chunk = next?;
+                c.write_all(&chunk)?;
+            }
+            c.set_position(0);
+            let mut xml = String::new();
+            c.read_to_string(&mut xml)?;
+            Bmap::from_xml(&xml)?
+        }
+    };
+
+    match input_type {
+        Input::Local(mut input) => {
+            let mut output = std::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(c.dest)?;
+
+            if output.metadata()?.is_file() {
+                ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
+                    .context("Failed to truncate file")?;
+            }
+            bmap::copy(&mut input, &mut output, &bmap)?;
+            println!("Done: Syncing...");
+            output.sync_all().expect("Sync failure");
+        }
+        Input::Remote(res) => {
+            let mut output = tokio::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(c.dest)
+                .await?;
+            ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
+                .context("Failed to truncate file")?;
```

Comment:
This should be done outside of the input_type determination as it's not dependant on the input